doc/man/man3/AppArgs.3
.TH "AppArgs" 3 "Sat Apr 7 2012" "Version 0.5.10" "incron" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AppArgs \- 
.PP
Application arguments'\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <appargs\&.h>\fP
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBInit\fP (size_t valMinCnt=0, size_t valMaxCnt=APPARGS_NOLIMIT)"
.br
.RI "\fIInitializes the processor'\&. \fP"
.ti -1c
.RI "static void \fBDestroy\fP ()"
.br
.RI "\fIReleases resources allocated by the processor'\&. \fP"
.ti -1c
.RI "static void \fBParse\fP (int argc, const char *const *argv)"
.br
.RI "\fIParses arguments and builds the appropriate structure'\&. \fP"
.ti -1c
.RI "static bool \fBIsValid\fP ()"
.br
.RI "\fIChecks whether the arguments have valid form'\&. \fP"
.ti -1c
.RI "static bool \fBExistsOption\fP (const std::string &rArg)"
.br
.RI "\fIChecks whether an option exists'\&. \fP"
.ti -1c
.RI "static bool \fBGetOption\fP (const std::string &rArg, std::string &rVal)"
.br
.RI "\fIExtracts an option value'\&. \fP"
.ti -1c
.RI "static bool \fBAddOption\fP (const std::string &rName, char cShort, \fBAppArgType_t\fP type, bool fMandatory)"
.br
.RI "\fIAdds an option'\&. \fP"
.ti -1c
.RI "static size_t \fBGetValueCount\fP ()"
.br
.RI "\fIReturns the count of values'\&. \fP"
.ti -1c
.RI "static bool \fBGetValue\fP (size_t index, std::string &rVal)"
.br
.RI "\fIExtracts a value'\&. \fP"
.ti -1c
.RI "static void \fBDump\fP ()"
.br
.RI "\fIDumps information about options and value to STDERR'\&. \fP"
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static bool \fBIsOption\fP (const char *pchStr)"
.br
.RI "\fIChecks whether a string is an option'\&. \fP"
.ti -1c
.RI "static bool \fBIsLongOption\fP (const char *pchStr)"
.br
.RI "\fIChecks whether a string is a long option'\&. \fP"
.ti -1c
.RI "static bool \fBParseLong\fP (const char *pchStr, std::string &rName, std::string &rVal, bool &rfHasVal)"
.br
.RI "\fIParses a string and attempts to treat it as a long option'\&. \fP"
.ti -1c
.RI "static void \fBParseShort\fP (const char *pchStr, char &rcName, std::string &rVal, bool &rfHasVal)"
.br
.RI "\fIParses a string and attempts to treat it as a short option'\&. \fP"
.ti -1c
.RI "static void \fBDumpOption\fP (const std::string &rName, char cShort, \fBAppArgOption_t\fP *pOpt)"
.br
.RI "\fIDumps an option to STDERR'\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
Application arguments'\&. 

This class is set-up for processing command line arguments'\&. Then it parses these arguments and builds data which can be queried later'\&.
.PP
There are two categories of arguments: 
.PD 0

.IP "\(bu" 2
options (a'\&.k'\&.a'\&. switches) 
.IP "\(bu" 2
values
.PP
Options represent changeable parameters of the application'\&. Values are a kind of input data'\&.
.PP
Each option has one of the following types: 
.PD 0

.IP "\(bu" 2
no value (two-state logic, e'\&.g'\&. running on foreground/background) 
.IP "\(bu" 2
optional value (e'\&.g'\&. for logging: another file than default can be specified) 
.IP "\(bu" 2
mandatory value (e'\&.g'\&. custom configuration file)
.PP
Each option always have two forms - long one (introcuded by two hyphens, e'\&.g'\&. --edit) and short one (introduced by one hyphen, e'\&.g'\&. -e)'\&. These forms are functionally equivalent'\&.
.PP
Unknown options are silently ignored'\&. 
.SH "Member Function Documentation"
.PP 
.SS "bool AppArgs::AddOption (const std::string &rName, charcShort, \fBAppArgType_t\fPtype, boolfMandatory)\fC [static]\fP"
.PP
Adds an option'\&. This method is intended to be called between initilization and parsing'\&. It adds an option which may (or must) occur inside the argument vector'\&.
.PP
\fBParameters:\fP
.RS 4
\fIrName\fP long option name 
.br
\fIcShort\fP short (one-character) option name 
.br
\fItype\fP argument type 
.br
\fIfMandatory\fP option is mandatory yes/no 
.RE
.PP
\fBReturns:\fP
.RS 4
true = success, false = failure (e'\&.g'\&. option already exists) 
.RE
.PP

.SS "void AppArgs::Destroy ()\fC [static]\fP"
.PP
Releases resources allocated by the processor'\&. This method should be called if the argument values are no longer needed'\&. 
.SS "void AppArgs::Dump ()\fC [static]\fP"
.PP
Dumps information about options and value to STDERR'\&. \fBAttention:\fP
.RS 4
This method may be very slow'\&. 
.RE
.PP

.SS "void AppArgs::DumpOption (const std::string &rName, charcShort, \fBAppArgOption_t\fP *pOpt)\fC [static, protected]\fP"
.PP
Dumps an option to STDERR'\&. \fBParameters:\fP
.RS 4
\fIrName\fP long option name 
.br
\fIcShort\fP short option name 
.br
\fIpOpt\fP option data 
.RE
.PP

.SS "bool AppArgs::ExistsOption (const std::string &rArg)\fC [static]\fP"
.PP
Checks whether an option exists'\&. \fBParameters:\fP
.RS 4
\fIrArg\fP long option name 
.RE
.PP
\fBReturns:\fP
.RS 4
true = option exists, false = otherwise 
.RE
.PP

.SS "bool AppArgs::GetOption (const std::string &rArg, std::string &rVal)\fC [static]\fP"
.PP
Extracts an option value'\&. \fBParameters:\fP
.RS 4
\fIrArg\fP long option name 
.br
\fIrVal\fP option value 
.RE
.PP
\fBReturns:\fP
.RS 4
true = value extracted, false = option not found or has no value 
.RE
.PP

.SS "bool AppArgs::GetValue (size_tindex, std::string &rVal)\fC [static]\fP"
.PP
Extracts a value'\&. \fBParameters:\fP
.RS 4
\fIindex\fP value index 
.br
\fIrVal\fP extracted value 
.RE
.PP
\fBReturns:\fP
.RS 4
true = value extracted, false = otherwise 
.RE
.PP

.SS "size_t AppArgs::GetValueCount ()\fC [static]\fP"
.PP
Returns the count of values'\&. \fBReturns:\fP
.RS 4
count of values 
.RE
.PP

.SS "void AppArgs::Init (size_tvalMinCnt = \fC0\fP, size_tvalMaxCnt = \fCAPPARGS_NOLIMIT\fP)\fC [static]\fP"
.PP
Initializes the processor'\&. \fBParameters:\fP
.RS 4
\fIvalMinCnt\fP minimum count of values 
.br
\fIvalMaxCnt\fP maximum number of values (no effect if lower than valMinCnt) 
.RE
.PP

.SS "bool AppArgs::IsLongOption (const char *pchStr)\fC [static, protected]\fP"
.PP
Checks whether a string is a long option'\&. This methos assumes the string is an option (if not the behavior is undefined)'\&.
.PP
\fBParameters:\fP
.RS 4
\fIpchStr\fP text string 
.RE
.PP
\fBReturns:\fP
.RS 4
true = option, false = otherwise 
.RE
.PP

.SS "bool AppArgs::IsOption (const char *pchStr)\fC [static, protected]\fP"
.PP
Checks whether a string is an option'\&. \fBParameters:\fP
.RS 4
\fIpchStr\fP text string 
.RE
.PP
\fBReturns:\fP
.RS 4
true = option, false = otherwise 
.RE
.PP

.SS "bool AppArgs::IsValid ()\fC [static]\fP"
.PP
Checks whether the arguments have valid form'\&. Arguments are valid if: 
.PD 0

.IP "\(bu" 2
all mandatory options are present 
.IP "\(bu" 2
all options with mandatory values have their values 
.IP "\(bu" 2
value count is between its minimum and maximum 
.IP "\(bu" 2
there are no unknown options (if unknown options are not accepted)
.PP
\fBReturns:\fP
.RS 4
true = arguments valid, false = otherwise 
.RE
.PP

.SS "void AppArgs::Parse (intargc, const char *const *argv)\fC [static]\fP"
.PP
Parses arguments and builds the appropriate structure'\&. \fBParameters:\fP
.RS 4
\fIargc\fP argument count 
.br
\fIargv\fP argument vector
.RE
.PP
\fBAttention:\fP
.RS 4
All errors are silently ignored'\&. 
.RE
.PP

.SS "bool AppArgs::ParseLong (const char *pchStr, std::string &rName, std::string &rVal, bool &rfHasVal)\fC [static, protected]\fP"
.PP
Parses a string and attempts to treat it as a long option'\&. \fBParameters:\fP
.RS 4
\fIpchStr\fP text string 
.br
\fIrName\fP option name 
.br
\fIrVal\fP value string 
.br
\fIrfHasVal\fP option has value yes/no 
.RE
.PP
\fBReturns:\fP
.RS 4
true = success, false = failure 
.RE
.PP

.SS "void AppArgs::ParseShort (const char *pchStr, char &rcName, std::string &rVal, bool &rfHasVal)\fC [static, protected]\fP"
.PP
Parses a string and attempts to treat it as a short option'\&. \fBParameters:\fP
.RS 4
\fIpchStr\fP text string 
.br
\fIrcName\fP option name 
.br
\fIrVal\fP value string 
.br
\fIrfHasVal\fP option has value yes/no
.RE
.PP
\fBAttention:\fP
.RS 4
This method assumes the string is a valid short option'\&. 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for incron from the source code'\&.


CHANGELOG
0.5.10   2012-04-05
* some includes fixed to build correctly with GCC 4.6 (and 4.4-4.5 - #0000385, #0000447)
* some type issues fixed (#0000447)
* EAGAIN handling in poll() (#0000399, #0000686, #0000691)
* Makefile improved as of #0000512
* typo in various files fixed (#0000412)


0.5.9    2009-06-21
* spaces in event-related file names now handled correctly (#0000174)
* crashing on 'incrontab --user' fixed (#0000187)
* man pages improved (bug reporting info as of #0000407)
* Makefile improved ('release-clean' also removes 'doc')


0.5.8    2008-12-29
* more includes added for GCC 4.3 compatibility (#0000178)
* typo in help messages fixed (#0000172)
* some warnings fixed (#0000372)


0.5.7    2007-09-04
* man pages updated (mainly due to #0000157)
* some includes added for GCC 4.3 compatibility (#0000158)
* supplementary groups initialization (#0000167)


0.5.6    2007-04-18
* based on inotify-cxx 0.7.2 (fixes compilation problems on Debian Sarge, #0000155)


0.5.5    2007-02-14
* another bug related to environment variables fixed (#0000151)
* -V (--version) malfunction fixed (#0000150)


0.5.4    2007-02-11
* bug related to persisting environment fixed (#0000149)
* incrond(8) man page updated to explain how the environment is treated
* buggy example configuration file fixed


0.5.3    2007-02-08
* bug related to ignoring new tables fixed (#0000142)
* explicit typecasting in execlp() call (#0000145)
* memory leak related to loaded tables fixed (#0000144)
* documentation for 'incrontab -d' added (#0000146)
* man page for incron.conf added (#0000147)


0.5.2    2007-02-04
* lockfile permissions changed to 0644 (reopened #0000137)
* incorrect reporting of removing non-existing tables fixed (#0000141)


0.5.1    2007-01-28
* lockfile location problems fixed (#0000137)
* lockfile configuration failure fixed (#0000140)
* CRLFs inside LICENCE-GPL changed to LFs (#0000138)


0.5.0    2007-01-27
* based on inotify-cxx 0.7.1
* bug related to events names fixed (#0000134)
* instance locking (only one instance allowed - also fixes #0000136)
* configuration introduced (including custom cfg. files)
* waiting for child processes rewritten to avoid race conditions
* user can request reloading the table (#0000130)
* new implementation for command line arguments
* some code refactoring


0.4.0    2007-01-13
* based on inotify-cxx 0.7.0
* incrontab has a feature to find out supported event types (--types or -t)
* incrond now supports command line arguments
* incrond can be run on foreground (--foreground, -n)
* incrond can be simply stopped (--kill, -k)
* support for system tables (/etc/incron.d) added
* eaccess() replaced by euiaccess() (#0000125)
* each table now use its own inotify descriptor (#0000129)
* descriptors are closed on executing child processes


0.3.5    2007-01-09
* based on inotify-cxx 0.6.3
* a security bug related to access rights to watched files fixed (#0000119)
* crashing on table reload after removing a watched path fixed (#0000123)
* 'incrontab --list' dysfunction fixed (#0000122)


0.3.4    2007-01-03
* based on inotify-cxx 0.6.2
* problems with wrong ordered setting of UID/GID fixed (#0000115, #0000117)
* user tables have 0600 permission now


0.3.3    2006-12-15
* based on inotify-cxx 0.5.3
* "GID ignorance" bug fixed (#0000109)
* "zombie bug" fixed (#0000105)
* code cleanup and simplifying


0.3.2    2006-12-02
* problems with reiserfs fixed (#0000104)


0.3.1    2006-11-24
* user manual added
* incron tables now owned by root (also fixes #0000103)


0.3.0    2006-11-12
* based on inotify-cxx 0.5.2 (fixes a problem with ignoring IN_OPEN)
* two new special symbols ($% for flags as names, $& for numeric flags)
* whitespaces in paths now allowed (by prepending a backslash) (#0000098)
* some changes in logging etc.


0.2.3    2006-10-30
* problems with unwanted IN_NO_LOOP have been fixed (#0000097)


0.2.2    2006-10-29
* based on inotify-cxx 0.5.0
* 'uncleaned children' bug has been fixed (#0000095)
* memory leak (child process arguments) has been fixed (#0000096)


0.2.1    2006-10-14
* based on inotify-cxx 0.4.1 (it should also fix crashing as described in #0000085)
* build failure (due to missing 'useradd -M') on some distributions has been fixed
  (#0000094)
* tarball structure has been fixed (#0000089)
* documentation slightly improved


0.2.0    2006-10-04
* based on inotify-cxx 0.3.1
* better signal handling (no "dirty" wait() in a handler)
* loop avoidance can be used
* more documentation in the code


0.1.0    2006-09-15
* first alpha version


appargs.cpp

/// application arguments processor implementation
/**
 * \file appargs.cpp
 * 
 * application arguments processor
 * 
 * Copyright (C) 2007, 2008, 2012 Lukas Jelinek, <lukas@aiken.cz>
 * Copyright (C) 2014, 2015 Andreas Altair Redmer, <altair.ibn.la.ahad.sy@gmail.com>
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of one of the following licenses:
 *
 * \li 1. X11-style license (see LICENSE-X11)
 * \li 2. GNU Lesser General Public License, version 2.1 (see LICENSE-LGPL)
 * \li 3. GNU General Public License, version 2  (see LICENSE-GPL)
 *
 * If you want to help with choosing the best license for you,
 * please visit http://www.gnu.org/licenses/license-list.html.
 * 
 * Credits:
 *   Christian Ruppert (new include to build with GCC 4.4+)
 * 
 */


#include <cstdio>
#include <cstring>

#include "strtok.h"

#include "appargs.h"


size_t AppArgs::s_minCnt = 0;
size_t AppArgs::s_maxCnt = APPARGS_NOLIMIT;

AA_LONG_MAP AppArgs::s_longMap;
AA_SHORT_MAP AppArgs::s_shortMap;
AA_VAL_LIST AppArgs::s_valList;


void AppArgs::Init(size_t valMinCnt, size_t valMaxCnt)
{
  s_minCnt = valMinCnt;
  s_maxCnt = valMaxCnt;
}

void AppArgs::Destroy()
{
  AA_LONG_MAP::iterator it = s_longMap.begin();
  while (it != s_longMap.end()) {
    delete (*it).second;
    it++;
  }
  
  s_longMap.clear();
  s_shortMap.clear();
  s_valList.clear();
}

void AppArgs::Parse(int argc, const char* const* argv)
{
  for (int i=1; i<argc; i++) {
    // this shouldn't occur
    if (argv[i] == NULL)
      return;
      
    if (IsOption(argv[i])) {
      if (IsLongOption(argv[i])) {
        std::string name, val;
        bool hasVal;
        if (ParseLong(argv[i], name, val, hasVal)) {
          AA_LONG_MAP::iterator it = s_longMap.find(name);
          if (it != s_longMap.end()) {
            AppArgOption_t* pOpt = (*it).second;
            pOpt->found = true;
            pOpt->hasVal = hasVal;
            pOpt->val = val;
          }
        }
      }
      else {
        char name;
        std::string val;
        bool hasVal;
        ParseShort(argv[i], name, val, hasVal);
        AA_SHORT_MAP::iterator it = s_shortMap.find(name);
        if (it != s_shortMap.end()) {
          AppArgOption_t* pOpt = (*it).second;
          pOpt->found = true;
          if (hasVal) {
            pOpt->hasVal = true;
            pOpt->val = val;
          }
          else {
            if (i+1 < argc && !IsOption(argv[i+1])) {
              pOpt->hasVal = true;
              pOpt->val = argv[i+1];
              i++;
            }
            else {
              pOpt->hasVal = false;
            }
          }
        }
      }
    }
    else {
      s_valList.push_back(argv[i]);
    }
  }
}

bool AppArgs::IsValid()
{
  size_t size = s_valList.size();
  if (size < s_minCnt || size > s_maxCnt)
    return false;
    
  AA_LONG_MAP::iterator it = s_longMap.begin();
  while (it != s_longMap.end()) {
    AppArgOption_t* pOpt = (*it).second;
    if (pOpt->mand && !(pOpt->found))
      return false;
    if (pOpt->type == AAT_MANDATORY_VALUE && pOpt->found && !(pOpt->hasVal))
      return false;
    it++;
  }
  
  return true;
}

bool AppArgs::ExistsOption(const std::string& rArg)
{
  AA_LONG_MAP::iterator it = s_longMap.find(rArg);
  return it != s_longMap.end() && (*it).second->found;
}

bool AppArgs::GetOption(const std::string& rArg, std::string& rVal)
{
  AA_LONG_MAP::iterator it = s_longMap.find(rArg);
  if (it == s_longMap.end())
    return false;
    
  AppArgOption_t* pOpt = (*it).second; 
  if (!(pOpt->found) || !(pOpt->hasVal))
    return false;
    
  rVal = pOpt->val;
  return true;
}

bool AppArgs::AddOption(const std::string& rName, char cShort, AppArgType_t type, bool fMandatory)
{
  if (s_longMap.find(rName) != s_longMap.end() || s_shortMap.find(cShort) != s_shortMap.end())
    return false;
  
  AppArgOption_t* pOpt = new AppArgOption_t();
  pOpt->found = false;
  pOpt->hasVal = false;
  pOpt->mand = fMandatory;
  pOpt->type = type;
  
  s_longMap.insert(AA_LONG_MAP::value_type(rName, pOpt));
  s_shortMap.insert(AA_SHORT_MAP::value_type(cShort, pOpt));
  return true;
}

size_t AppArgs::GetValueCount()
{
  return s_valList.size();
}

bool AppArgs::GetValue(size_t index, std::string& rVal)
{
  if (index > s_valList.size())
    return false;
    
  rVal = s_valList[index];
  return true;
}

void AppArgs::Dump()
{
  fprintf(stderr, "Options:\n");
  AA_LONG_MAP::iterator it = s_longMap.begin();
  while (it != s_longMap.end()) {
    AppArgOption_t* pOpt = (*it).second;
    AA_SHORT_MAP::iterator it2 = s_shortMap.begin();
    while (it2 != s_shortMap.end()) {
      if ((*it2).second == pOpt) {
        DumpOption((*it).first, (*it2).first, pOpt);
      }
      it2++;
    }      
    it++;
  }
  
  fprintf(stderr, "Values:\n");
  
  AA_VAL_LIST::iterator it3 = s_valList.begin();
  while (it3 != s_valList.end()) {
    fprintf(stderr, "%s\n", (*it3).c_str());
    it3++;
  }  
}



bool AppArgs::IsOption(const char* pchStr)
{
  if (strlen(pchStr) < 2)
    return false;
    
  return pchStr[0] == '-';
}

bool AppArgs::IsLongOption(const char* pchStr)
{
  return pchStr[1] == '-';
}

bool AppArgs::ParseLong(const char* pchStr, std::string& rName, std::string& rVal, bool& rfHasVal)
{
  StringTokenizer tok(pchStr+2, "=");
  if (!tok.HasMoreTokens())
    return false;
  
  rName = tok.GetNextToken();
  if (!tok.HasMoreTokens()) {
    rfHasVal = false;
    return true;
  }
    
  rVal = tok.GetRemainder();
  rfHasVal = true;
  return true;
}

void AppArgs::ParseShort(const char* pchStr, char& rcName, std::string& rVal, bool& rfHasVal)
{
  rcName = pchStr[1];
  size_t len = strlen(pchStr);
  if (len == 2) {
    rfHasVal = false;
    return;
  }
  
  rVal = pchStr + 2;
  rfHasVal = true;  
}

void AppArgs::DumpOption(const std::string& rName, char cShort, AppArgOption_t* pOpt)
{
  const char* s;
  switch (pOpt->type) {
    case AAT_NO_VALUE: s = "no value";
      break;
    case AAT_OPTIONAL_VALUE: s = "optional value";
      break;
    case AAT_MANDATORY_VALUE: s = "mandatory value";
      break;
    default:;
      s = "unknown";
  }
  
  fprintf(stderr, "long='%s', short='%c', type='%s', found=%s, has_value=%s, value='%s'\n",
      rName.c_str(),
      cShort,
      s,
      pOpt->found ? "YES" : "NO",
      pOpt->hasVal ? "YES" : "NO",
      pOpt->val.c_str()
  );
}


